# Spring 3 Tasca 2 

### Exercici 1

Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

Redueix la dimensió del dataset de manera aleatòria per tal d'obtenir un dataset de només 200.000 registres. Tots els exercicis s'han de fer amb aquest dataset reduït.

In [1]:
import pandas as pd

In [2]:
ruta_archivo = "C:\\Users\\gemma\\OneDrive\\Escritorio\\It Academy Tasks\\Archivo_ Spr3_aviones\\DelayedFlights.csv"

In [3]:
df = pd.read_csv(ruta_archivo , index_col=0)

In [4]:
df

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0        2008      1           3          4   2003.0        1955   2211.0   
1        2008      1           3          4    754.0         735   1002.0   
2        2008      1           3          4    628.0         620    804.0   
4        2008      1           3          4   1829.0        1755   1959.0   
5        2008      1           3          4   1940.0        1915   2121.0   
...       ...    ...         ...        ...      ...         ...      ...   
7009710  2008     12          13          6   1250.0        1220   1617.0   
7009717  2008     12          13          6    657.0         600    904.0   
7009718  2008     12          13          6   1007.0         847   1149.0   
7009726  2008     12          13          6   1251.0        1240   1446.0   
7009727  2008     12          13          6   1110.0        1103   1413.0   

         CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0              2225            WN        335  ...    4.0      8.0          0   
1              1000            WN       3231  ...    5.0     10.0          0   
2               750            WN        448  ...    3.0     17.0          0   
4              1925            WN       3920  ...    3.0     10.0          0   
5              2110            WN        378  ...    4.0     10.0          0   
...             ...           ...        ...  ...    ...      ...        ...   
7009710        1552            DL       1621  ...    9.0     18.0          0   
7009717         749            DL       1631  ...   15.0     34.0          0   
7009718        1010            DL       1631  ...    8.0     32.0          0   
7009726        1437            DL       1639  ...   13.0     13.0          0   
7009727        1418            DL       1641  ...    8.0     11.0          0   

         CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                       N         0           NaN          NaN      NaN   
1                       N         0           NaN          NaN      NaN   
2                       N         0           NaN          NaN      NaN   
4                       N         0           2.0          0.0      0.0   
5                       N         0           NaN          NaN      NaN   
...                   ...       ...           ...          ...      ...   
7009710                 N         0           3.0          0.0      0.0   
7009717                 N         0           0.0         57.0     18.0   
7009718                 N         0           1.0          0.0     19.0   
7009726                 N         0           NaN          NaN      NaN   
7009727                 N         0           NaN          NaN      NaN   

         SecurityDelay  LateAircraftDelay  
0                  NaN                NaN  
1                  NaN                NaN  
2                  NaN                NaN  
4                  0.0               32.0  
5                  NaN                NaN  
...                ...                ...  
7009710            0.0               22.0  
7009717            0.0                0.0  
7009718            0.0               79.0  
7009726            NaN                NaN  
7009727            NaN                NaN  

[1936758 rows x 29 columns]

El dataset està compost per les següents columnes:

    - Year (any): 2008
    - Month (mesos): 1-12
    - DayofMonth (dia del mes): 1-31
    - DayOfWeek 1 (Dilluns) - 7 (Diumenge)
    - DepTime (Hora de sortida real) (local, hhmm)
    - CRSDepTime (Hora de sortida prevista) (local, hhmm)
    - ArrTime (Hora d'arribada real) (local, hhmm)
    - CRSArrTime (Hora d'arribada prevista) (local, hhmm)
    - UniqueCarrier (Còdig identificador aerolinia)
    - FlightNum (número de vol)
    - TailNum (id de l'avió, és únic
    - ActualElapsedTime (duració del vol real): en minuts
    - CRSElapsedTime (duració del vol estimada): en minuts
    - AirTime en minuts
    - ArrDelay (arribada amb retard) en minuts. UEs considera que ha arribat a temps quan el retràs és inferior a 15 minuts.
    - DepDelay (minuts de retard de la sortida): en minuts
    - Origin (origen): còdig de l'aeroport  
    - Dest (destinació): còdig del destí
    - Distance (distancia) en milles
    - TaxiIn (taxi per arribar): en minuts
    - TaxiOut (taxi per marxar): en minuts
    - Cancelled *si el vol s'ha cancel3lat)
    - CancellationCode (còdig amb motiu de cancel·lació A = companyia, B = temps, C = NAS, D = seguretat)
    - Diverted (vol desviat) 1 = si, 0 = no
    - CarrierDelay (retràs ocasionat per la Companyia): en minuts
    - WeatherDelay: (retràs ocasiont pel temps): en minuts
    - NASDelay (retàs ocasionat per National Airspace System (NAS)): en minuts
    - SecurityDelay (retràs ocasionat per Seguretat): en minuts
    - LateAircraftDelay (arribada a l'aeroport d'origen amb retràs - venint d'una altre vol): en minuts
    
    
    

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1936758 entries, 0 to 7009727
Data columns (total 29 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Year               int64  
 1   Month              int64  
 2   DayofMonth         int64  
 3   DayOfWeek          int64  
 4   DepTime            float64
 5   CRSDepTime         int64  
 6   ArrTime            float64
 7   CRSArrTime         int64  
 8   UniqueCarrier      object 
 9   FlightNum          int64  
 10  TailNum            object 
 11  ActualElapsedTime  float64
 12  CRSElapsedTime     float64
 13  AirTime            float64
 14  ArrDelay           float64
 15  DepDelay           float64
 16  Origin             object 
 17  Dest               object 
 18  Distance           int64  
 19  TaxiIn             float64
 20  TaxiOut            float64
 21  Cancelled          int64  
 22  CancellationCode   object 
 23  Diverted           int64  
 24  CarrierDelay       float64
 25  WeatherDelay      

Primer eliminarem duplicats

In [6]:
df.drop_duplicates()
print(df.shape)

(1936758, 29)


En aquest cas no hi havia duplicats.

Abans de començar a analitzar les dades és important saber què volem analitzar i amb quina finalitat. Ja que això ens permetrà decidir quines columnes i/o registres hem d'eliminar. L'objectiu del meu anàlisis és estudiar més detingudament la quantitat de retards que hi han hagut respecte a tots els vols que s'han dut a terme i que han arribar al seu destí. Poder valorar de quins són els aeroports i les companyies aerees amb més retards. Així com veure quines són les causes que més els provoquen.

Per a fer això el primer pas és eliminar tots aquells registres que tinguin el vol cancel.lat o que hagin estat desviats a un altre destí.

In [7]:
# Els valor dels vols cancel·lats i desviats és 1. El valor per a la resta de vols és 0.

In [8]:
new_df = df[(df["Cancelled"] != 1) & (df["Diverted"] != 1)]
print(new_df.shape)

(1928371, 29)


Un cop eliminats els registres de vols cancel·lats i vols deviats, eliminaré les columnes que no considero necessàries per al meu anàlisis:
- Year: totes les dades són de 2008
- TaxiIn
- TaxiOut
- Cancelled
- CancellationCode
- Diverted

In [9]:
new_df_2=new_df.drop(["Year","TaxiIn","TaxiOut","Cancelled","CancellationCode","Diverted"],axis=1)
print(new_df_2.shape)

(1928371, 23)


Un cop eliminades les columnes que ja no son necessàries, el següent pas es reduir el dataset a 200.000 registres.

In [10]:
df_reduced = new_df_2.sample(n=200000,random_state=27)
#he escollit la prova 27, i s'indica per si algú més volfer servir la mateixa mostra

In [11]:
df_reduced


Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
3528691      6          11          3   1145.0        1126   1325.0   
3796307      7          11          5   1308.0        1300   1528.0   
4474620      8          16          6   2151.0        2050     38.0   
6833490     12           1          1   2019.0        1835   2125.0   
5378051      9          10          3   2200.0        2140   2251.0   
...        ...         ...        ...      ...         ...      ...   
2970098      5          27          2   1914.0        1725   2350.0   
3662033      7          20          7   1901.0        1850   2338.0   
1127926      2          19          2   2115.0        1945   2239.0   
4482356      8           7          4   1056.0        1021   1232.0   
3707690      7          15          2   2013.0        1940   2141.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ...  ArrDelay  DepDelay  \
3528691        1313            AS        366  N553AS  ...      12.0      19.0   
3796307        1522            OO       5880  N941SW  ...       6.0       8.0   
4474620        2337            UA        628  N414UA  ...      61.0      61.0   
6833490        1930            MQ       3956  N648MQ  ...     115.0     104.0   
5378051        2229            DL        965  N930DL  ...      22.0      20.0   
...             ...           ...        ...     ...  ...       ...       ...   
2970098        2025            CO       1426  N33203  ...     205.0     109.0   
3662033        2345            WN        621  N474WN  ...      -7.0      11.0   
1127926        2110            B6         33  N267JB  ...      89.0      90.0   
4482356        1154            UA        900  N924UA  ...      38.0      35.0   
3707690        2113            XE       7771  N16170  ...      28.0      33.0   

         Origin  Dest  Distance CarrierDelay WeatherDelay  NASDelay  \
3528691     SEA   SMF       605          NaN          NaN       NaN   
3796307     DEN   LNK       423          NaN          NaN       NaN   
4474620     ORD   DCA       612         21.0          0.0       0.0   
6833490     ORD   GRB       174        104.0          0.0      11.0   
5378051     ATL   STL       483         12.0          0.0       2.0   
...         ...   ...       ...          ...          ...       ...   
2970098     EWR   SAN      2425          0.0          0.0     205.0   
3662033     LAS   MCI      1140          NaN          NaN       NaN   
1127926     ROC   JFK       264         89.0          0.0       0.0   
4482356     SAN   SFO       447          0.0          0.0      38.0   
3707690     RNO   LAX       390          0.0          0.0       0.0   

         SecurityDelay  LateAircraftDelay  
3528691            NaN                NaN  
3796307            NaN                NaN  
4474620            0.0               40.0  
6833490            0.0                0.0  
5378051            0.0                8.0  
...                ...                ...  
2970098            0.0                0.0  
3662033            NaN                NaN  
1127926            0.0                0.0  
4482356            0.0                0.0  
3707690            0.0               28.0  

[200000 rows x 23 columns]

### Exercici 2

Fes un informe complet del dataset:


- Troba quantes dades faltants hi ha per columna.

Mirarem si hi ha columnes amb sense valors.

In [12]:
df_reduced.isna().sum()

Month                    0
DayofMonth               0
DayOfWeek                0
DepTime                  0
CRSDepTime               0
ArrTime                  0
CRSArrTime               0
UniqueCarrier            0
FlightNum                0
TailNum                  1
ActualElapsedTime        0
CRSElapsedTime           0
AirTime                  0
ArrDelay                 0
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
CarrierDelay         70622
WeatherDelay         70622
NASDelay             70622
SecurityDelay        70622
LateAircraftDelay    70622
dtype: int64

Podem observar que el dataset no té registres buits.

El següent pas es mirar si hi ha valors null a les columnes.

In [13]:
df_reduced.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200000 entries, 3528691 to 3707690
Data columns (total 23 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Month              200000 non-null  int64  
 1   DayofMonth         200000 non-null  int64  
 2   DayOfWeek          200000 non-null  int64  
 3   DepTime            200000 non-null  float64
 4   CRSDepTime         200000 non-null  int64  
 5   ArrTime            200000 non-null  float64
 6   CRSArrTime         200000 non-null  int64  
 7   UniqueCarrier      200000 non-null  object 
 8   FlightNum          200000 non-null  int64  
 9   TailNum            199999 non-null  object 
 10  ActualElapsedTime  200000 non-null  float64
 11  CRSElapsedTime     200000 non-null  float64
 12  AirTime            200000 non-null  float64
 13  ArrDelay           200000 non-null  float64
 14  DepDelay           200000 non-null  float64
 15  Origin             200000 non-null  object 


S'observa que el dataset té valors nulls. De manera que s'ha de modificar aquests valors. Observant la base de dades he vist que els valors NaN estàn en el les columnes relacionades amb el valor en minuts de les diferents causes de retards i són dels avions que no han tingut retards. De manera que podrem substituir sense problemes Nan per 0.0.

In [14]:
df_reduced.fillna(0.0, inplace = True)
df_reduced.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200000 entries, 3528691 to 3707690
Data columns (total 23 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Month              200000 non-null  int64  
 1   DayofMonth         200000 non-null  int64  
 2   DayOfWeek          200000 non-null  int64  
 3   DepTime            200000 non-null  float64
 4   CRSDepTime         200000 non-null  int64  
 5   ArrTime            200000 non-null  float64
 6   CRSArrTime         200000 non-null  int64  
 7   UniqueCarrier      200000 non-null  object 
 8   FlightNum          200000 non-null  int64  
 9   TailNum            200000 non-null  object 
 10  ActualElapsedTime  200000 non-null  float64
 11  CRSElapsedTime     200000 non-null  float64
 12  AirTime            200000 non-null  float64
 13  ArrDelay           200000 non-null  float64
 14  DepDelay           200000 non-null  float64
 15  Origin             200000 non-null  object 


- Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.

El primer que farem serà fer un breu resum estadistic en el que es mostrarà el total, la mitja, la desviació std, el valor mínim i el valor màxim així com també els percentils.
Aquesta informació és útil per poder veure per sobre la dispersió de les dades i si hi ha algun outliers. Especialment en determinades columnes com AirTime, Distance, i la informació relacionada amb els delays.

In [15]:
df_reduced.describe().T

count         mean          std   min     25%     50%  \
Month              200000.0     6.096165     3.482512   1.0     3.0     6.0   
DayofMonth         200000.0    15.751675     8.792437   1.0     8.0    16.0   
DayOfWeek          200000.0     3.986390     1.998295   1.0     2.0     4.0   
DepTime            200000.0  1518.905110   450.480212   1.0  1204.0  1546.0   
CRSDepTime         200000.0  1467.587595   424.526505   1.0  1135.0  1510.0   
ArrTime            200000.0  1611.134920   547.870956   1.0  1317.0  1716.0   
CRSArrTime         200000.0  1635.460955   463.782288   0.0  1325.0  1708.0   
FlightNum          200000.0  2191.777430  1946.686054   1.0   614.0  1551.0   
ActualElapsedTime  200000.0   133.043600    71.687593  15.0    80.0   116.0   
CRSElapsedTime     200000.0   133.986260    70.925915   1.0    82.0   116.0   
AirTime            200000.0   108.034565    68.283306   1.0    58.0    90.0   
ArrDelay           200000.0    42.280065    57.191270 -67.0     9.0    25.0   
DepDelay           200000.0    43.222715    53.748809   6.0    12.0    24.0   
Distance           200000.0   763.299920   571.718302  30.0   338.0   606.0   
CarrierDelay       200000.0    12.466800    36.925619   0.0     0.0     0.0   
WeatherDelay       200000.0     2.355140    17.090555   0.0     0.0     0.0   
NASDelay           200000.0     9.770975    28.491180   0.0     0.0     0.0   
SecurityDelay      200000.0     0.059575     1.654482   0.0     0.0     0.0   
LateAircraftDelay  200000.0    16.389100    35.804586   0.0     0.0     0.0   

                      75%     max  
Month                 9.0    12.0  
DayofMonth           23.0    31.0  
DayOfWeek             6.0     7.0  
DepTime            1901.0  2400.0  
CRSDepTime         1816.0  2359.0  
ArrTime            2031.0  2400.0  
CRSArrTime         2014.0  2359.0  
FlightNum          3442.0  9740.0  
ActualElapsedTime   164.0   776.0  
CRSElapsedTime      165.0   660.0  
AirTime             137.0   648.0  
ArrDelay             56.0  1370.0  
DepDelay             53.0  1521.0  
Distance            992.0  4962.0  
CarrierDelay         10.0  1348.0  
WeatherDelay          0.0  1153.0  
NASDelay              6.0  1289.0  
SecurityDelay         0.0   199.0  
LateAircraftDelay    18.0   925.0

D'aquesta taula podem treure informació important com que més d'un 25% dels vols arriben a temps i un 50% dels vols han arribat com a mínim amb 50 minuts de retards.Hi ha més vols que han sortit amb retards, però una part recupera el temps durant el vol.

Per tal de poder veure la relació entre les diferents variables. Fem una correlació entre elles.

In [16]:
df_reduced.corr()

Month  DayofMonth  DayOfWeek   DepTime  CRSDepTime  \
Month              1.000000    0.061245   0.001685 -0.007341   -0.011027   
DayofMonth         0.061245    1.000000   0.015950  0.002475    0.002803   
DayOfWeek          0.001685    0.015950   1.000000  0.022629    0.026947   
DepTime           -0.007341    0.002475   0.022629  1.000000    0.883303   
CRSDepTime        -0.011027    0.002803   0.026947  0.883303    1.000000   
ArrTime           -0.000955    0.001945   0.009626  0.461112    0.400441   
CRSArrTime         0.001481    0.003510   0.017930  0.716121    0.715511   
FlightNum         -0.002585   -0.005996  -0.011185 -0.024060   -0.053948   
ActualElapsedTime  0.001384    0.000078   0.004053 -0.049050   -0.035963   
CRSElapsedTime     0.006370    0.000813   0.005561 -0.046459   -0.027418   
AirTime           -0.000146    0.000132   0.005899 -0.057450   -0.038290   
ArrDelay           0.000565    0.003556   0.005604  0.127993    0.047994   
DepDelay           0.007161    0.004753   0.007895  0.140305    0.062853   
Distance           0.005431    0.001000   0.009343 -0.058149   -0.030800   
CarrierDelay      -0.001720   -0.002072   0.005914 -0.008211   -0.061326   
WeatherDelay       0.003105    0.000894   0.003805  0.012836    0.001060   
NASDelay           0.004080    0.003236  -0.003650  0.048339   -0.008345   
SecurityDelay     -0.001065   -0.000418   0.003833 -0.008182   -0.010718   
LateAircraftDelay -0.000392    0.004527   0.003804  0.172960    0.148839   

                    ArrTime  CRSArrTime  FlightNum  ActualElapsedTime  \
Month             -0.000955    0.001481  -0.002585           0.001384   
DayofMonth         0.001945    0.003510  -0.005996           0.000078   
DayOfWeek          0.009626    0.017930  -0.011185           0.004053   
DepTime            0.461112    0.716121  -0.024060          -0.049050   
CRSDepTime         0.400441    0.715511  -0.053948          -0.035963   
ArrTime            1.000000    0.618035  -0.015025          -0.014692   
CRSArrTime         0.618035    1.000000  -0.061059           0.032567   
FlightNum         -0.015025   -0.061059   1.000000          -0.320475   
ActualElapsedTime -0.014692    0.032567  -0.320475           1.000000   
CRSElapsedTime    -0.014270    0.039242  -0.333393           0.971158   
AirTime           -0.019646    0.024768  -0.339213           0.976656   
ArrDelay          -0.051317    0.045230   0.058398           0.069779   
DepDelay          -0.053839    0.056473   0.049635           0.022018   
Distance          -0.029265    0.023897  -0.354646           0.952742   
CarrierDelay      -0.062409   -0.057338   0.058508           0.016264   
WeatherDelay      -0.029274    0.002504   0.061621           0.007431   
NASDelay           0.032092    0.021262   0.027444           0.147105   
SecurityDelay     -0.006390   -0.010275   0.002897           0.004691   
LateAircraftDelay -0.027182    0.117503  -0.020886          -0.008927   

                   CRSElapsedTime   AirTime  ArrDelay  DepDelay  Distance  \
Month                    0.006370 -0.000146  0.000565  0.007161  0.005431   
DayofMonth               0.000813  0.000132  0.003556  0.004753  0.001000   
DayOfWeek                0.005561  0.005899  0.005604  0.007895  0.009343   
DepTime                 -0.046459 -0.057450  0.127993  0.140305 -0.058149   
CRSDepTime              -0.027418 -0.038290  0.047994  0.062853 -0.030800   
ArrTime                 -0.014270 -0.019646 -0.051317 -0.053839 -0.029265   
CRSArrTime               0.039242  0.024768  0.045230  0.056473  0.023897   
FlightNum               -0.333393 -0.339213  0.058398  0.049635 -0.354646   
ActualElapsedTime        0.971158  0.976656  0.069779  0.022018  0.952742   
CRSElapsedTime           1.000000  0.985991 -0.012771  0.010708  0.981761   
AirTime                  0.985991  1.000000  0.001862  0.000460  0.980086   
ArrDelay                -0.012771  0.001862  1.000000  0.954127 -0.027413   
DepDelay                 0.010708  0.00

- Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).

Crearem una columna anomenda Delay que determini si un avió ha arribat tard o no. Els valors que tindrà aquesta columna són:
    "Yes" = ArrDelay <= 15 (minuts)
    "No" = ArrDelay > 15 (minuts)

In [17]:
def is_delay(ArrDelay):
    if ArrDelay > 15:
        return "Yes"
    else:
        return "No"
    
df_reduced["Delay"] = df_reduced["ArrDelay"].apply (is_delay)
df_reduced

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
3528691      6          11          3   1145.0        1126   1325.0   
3796307      7          11          5   1308.0        1300   1528.0   
4474620      8          16          6   2151.0        2050     38.0   
6833490     12           1          1   2019.0        1835   2125.0   
5378051      9          10          3   2200.0        2140   2251.0   
...        ...         ...        ...      ...         ...      ...   
2970098      5          27          2   1914.0        1725   2350.0   
3662033      7          20          7   1901.0        1850   2338.0   
1127926      2          19          2   2115.0        1945   2239.0   
4482356      8           7          4   1056.0        1021   1232.0   
3707690      7          15          2   2013.0        1940   2141.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ...  DepDelay  Origin  \
3528691        1313            AS        366  N553AS  ...      19.0     SEA   
3796307        1522            OO       5880  N941SW  ...       8.0     DEN   
4474620        2337            UA        628  N414UA  ...      61.0     ORD   
6833490        1930            MQ       3956  N648MQ  ...     104.0     ORD   
5378051        2229            DL        965  N930DL  ...      20.0     ATL   
...             ...           ...        ...     ...  ...       ...     ...   
2970098        2025            CO       1426  N33203  ...     109.0     EWR   
3662033        2345            WN        621  N474WN  ...      11.0     LAS   
1127926        2110            B6         33  N267JB  ...      90.0     ROC   
4482356        1154            UA        900  N924UA  ...      35.0     SAN   
3707690        2113            XE       7771  N16170  ...      33.0     RNO   

         Dest  Distance  CarrierDelay WeatherDelay NASDelay  SecurityDelay  \
3528691   SMF       605           0.0          0.0      0.0            0.0   
3796307   LNK       423           0.0          0.0      0.0            0.0   
4474620   DCA       612          21.0          0.0      0.0            0.0   
6833490   GRB       174         104.0          0.0     11.0            0.0   
5378051   STL       483          12.0          0.0      2.0            0.0   
...       ...       ...           ...          ...      ...            ...   
2970098   SAN      2425           0.0          0.0    205.0            0.0   
3662033   MCI      1140           0.0          0.0      0.0            0.0   
1127926   JFK       264          89.0          0.0      0.0            0.0   
4482356   SFO       447           0.0          0.0     38.0            0.0   
3707690   LAX       390           0.0          0.0      0.0            0.0   

         LateAircraftDelay  Delay  
3528691                0.0     No  
3796307                0.0     No  
4474620               40.0    Yes  
6833490                0.0    Yes  
5378051                8.0    Yes  
...                    ...    ...  
2970098                0.0    Yes  
3662033                0.0     No  
1127926                0.0    Yes  
4482356                0.0    Yes  
3707690               28.0    Yes  

[200000 rows x 24 columns]

Crearem una nova columna que calcularà els minuts de retard de la sortida dels avions.

Per poder estudiar més detingudament els retards crearem un subset en el que només apareguin els vols amb un retard a l'arrivada superior a 15 minuts.

In [18]:
df_reduced["DepDelMin"]= df_reduced["DepTime"]- df_reduced["CRSDepTime"]
df_reduced

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
3528691      6          11          3   1145.0        1126   1325.0   
3796307      7          11          5   1308.0        1300   1528.0   
4474620      8          16          6   2151.0        2050     38.0   
6833490     12           1          1   2019.0        1835   2125.0   
5378051      9          10          3   2200.0        2140   2251.0   
...        ...         ...        ...      ...         ...      ...   
2970098      5          27          2   1914.0        1725   2350.0   
3662033      7          20          7   1901.0        1850   2338.0   
1127926      2          19          2   2115.0        1945   2239.0   
4482356      8           7          4   1056.0        1021   1232.0   
3707690      7          15          2   2013.0        1940   2141.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ...  Origin  Dest  \
3528691        1313            AS        366  N553AS  ...     SEA   SMF   
3796307        1522            OO       5880  N941SW  ...     DEN   LNK   
4474620        2337            UA        628  N414UA  ...     ORD   DCA   
6833490        1930            MQ       3956  N648MQ  ...     ORD   GRB   
5378051        2229            DL        965  N930DL  ...     ATL   STL   
...             ...           ...        ...     ...  ...     ...   ...   
2970098        2025            CO       1426  N33203  ...     EWR   SAN   
3662033        2345            WN        621  N474WN  ...     LAS   MCI   
1127926        2110            B6         33  N267JB  ...     ROC   JFK   
4482356        1154            UA        900  N924UA  ...     SAN   SFO   
3707690        2113            XE       7771  N16170  ...     RNO   LAX   

         Distance  CarrierDelay  WeatherDelay NASDelay SecurityDelay  \
3528691       605           0.0           0.0      0.0           0.0   
3796307       423           0.0           0.0      0.0           0.0   
4474620       612          21.0           0.0      0.0           0.0   
6833490       174         104.0           0.0     11.0           0.0   
5378051       483          12.0           0.0      2.0           0.0   
...           ...           ...           ...      ...           ...   
2970098      2425           0.0           0.0    205.0           0.0   
3662033      1140           0.0           0.0      0.0           0.0   
1127926       264          89.0           0.0      0.0           0.0   
4482356       447           0.0           0.0     38.0           0.0   
3707690       390           0.0           0.0      0.0           0.0   

         LateAircraftDelay  Delay  DepDelMin  
3528691                0.0     No       19.0  
3796307                0.0     No        8.0  
4474620               40.0    Yes      101.0  
6833490                0.0    Yes      184.0  
5378051                8.0    Yes       60.0  
...                    ...    ...        ...  
2970098                0.0    Yes      189.0  
3662033                0.0     No       51.0  
1127926                0.0    Yes      170.0  
4482356                0.0    Yes       35.0  
3707690               28.0    Yes       73.0  

[200000 rows x 25 columns]

Crearem una columna que calculi la mitjana de velocitat de cada vol. La columna es dirà Velocity.

In [19]:
df_reduced["Velocity"] = round(df_reduced["Distance"] / df_reduced["AirTime"],2)
df_reduced

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
3528691      6          11          3   1145.0        1126   1325.0   
3796307      7          11          5   1308.0        1300   1528.0   
4474620      8          16          6   2151.0        2050     38.0   
6833490     12           1          1   2019.0        1835   2125.0   
5378051      9          10          3   2200.0        2140   2251.0   
...        ...         ...        ...      ...         ...      ...   
2970098      5          27          2   1914.0        1725   2350.0   
3662033      7          20          7   1901.0        1850   2338.0   
1127926      2          19          2   2115.0        1945   2239.0   
4482356      8           7          4   1056.0        1021   1232.0   
3707690      7          15          2   2013.0        1940   2141.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ...  Dest  Distance  \
3528691        1313            AS        366  N553AS  ...   SMF       605   
3796307        1522            OO       5880  N941SW  ...   LNK       423   
4474620        2337            UA        628  N414UA  ...   DCA       612   
6833490        1930            MQ       3956  N648MQ  ...   GRB       174   
5378051        2229            DL        965  N930DL  ...   STL       483   
...             ...           ...        ...     ...  ...   ...       ...   
2970098        2025            CO       1426  N33203  ...   SAN      2425   
3662033        2345            WN        621  N474WN  ...   MCI      1140   
1127926        2110            B6         33  N267JB  ...   JFK       264   
4482356        1154            UA        900  N924UA  ...   SFO       447   
3707690        2113            XE       7771  N16170  ...   LAX       390   

         CarrierDelay  WeatherDelay  NASDelay SecurityDelay LateAircraftDelay  \
3528691           0.0           0.0       0.0           0.0               0.0   
3796307           0.0           0.0       0.0           0.0               0.0   
4474620          21.0           0.0       0.0           0.0              40.0   
6833490         104.0           0.0      11.0           0.0               0.0   
5378051          12.0           0.0       2.0           0.0               8.0   
...               ...           ...       ...           ...               ...   
2970098           0.0           0.0     205.0           0.0               0.0   
3662033           0.0           0.0       0.0           0.0               0.0   
1127926          89.0           0.0       0.0           0.0               0.0   
4482356           0.0           0.0      38.0           0.0               0.0   
3707690           0.0           0.0       0.0           0.0              28.0   

         Delay  DepDelMin  Velocity  
3528691     No       19.0      8.07  
3796307     No        8.0      7.55  
4474620    Yes      101.0      7.20  
6833490    Yes      184.0      5.61  
5378051    Yes       60.0      5.96  
...        ...        ...       ...  
2970098    Yes      189.0      6.91  
3662033     No       51.0      8.09  
1127926    Yes      170.0      3.67  
4482356    Yes       35.0      6.21  
3707690    Yes       73.0      5.82  

[200000 rows x 26 columns]

- Fes una taula de les aerolínies amb més endarreriments acumulats.

In [20]:
tabla_aerolineas = pd.crosstab(df_reduced["UniqueCarrier"],df_reduced["Delay"])
tabla_aerolineas["TotalFlights"] = tabla_aerolineas.sum(axis=1)
tabla_aerolineas["PercYes"] = round(tabla_aerolineas["Yes"] /tabla_aerolineas["TotalFlights"],2)
tabla_aerolineas_ordenado = tabla_aerolineas.sort_values(by="PercYes", ascending=False)
tabla_aerolineas_ordenado



Delay             No    Yes  TotalFlights  PercYes
UniqueCarrier                                     
YV              1774   5198          6972     0.75
OH              1502   4089          5591     0.73
NW              2544   5607          8151     0.69
AA              6342  13622         19964     0.68
XE              3421   7310         10731     0.68
9E              1751   3683          5434     0.68
B6              1795   3871          5666     0.68
EV              2741   5572          8313     0.67
MQ              4837   9824         14661     0.67
OO              4644   9042         13686     0.66
UA              4958   9438         14396     0.66
FL              2557   4740          7297     0.65
DL              4381   7382         11763     0.63
AS              1580   2368          3948     0.60
US              4184   5923         10107     0.59
HA               342    468           810     0.58
CO              4439   5814         10253     0.57
F9              1337   1542          2879     0.54
WN             18726  20587         39313     0.52
AQ                36     29            65     0.45

En la taula d'Aerolinees s'ha calculat el total de vols així com el % de vols amb retards. Per saber quina és la Aerolinea que més vols amb retard té en relació amb el total dels seus vols.  Observem que les companyes YV i OH són les que tenen una proporció de vols amb més retards. Però és destacable que de les 20 Aerolinies que hi ha només 1 està per sota del 50% de retards. Però nomès ha operat 65 vols.

- Quins són els vols més llargs? 

In [21]:
df_reduced.columns
taula_vols = df_reduced[["UniqueCarrier", "FlightNum", "TailNum", "AirTime","Origin","Dest","Distance", "Velocity","ArrDelay"]]
taula_vols_ord = taula_vols.sort_values(by="Distance", ascending=False)
taula_vols_ord.head(50)

UniqueCarrier  FlightNum TailNum  AirTime Origin Dest  Distance  \
5909296            CO         15  N68061    611.0    EWR  HNL      4962   
1742551            CO         15  N66057    622.0    EWR  HNL      4962   
4808097            CO         15  N59053    598.0    EWR  HNL      4962   
5906279            CO         15  N76062    584.0    EWR  HNL      4962   
1147136            CO         15  N66057    620.0    EWR  HNL      4962   
1742548            CO         15  N66056    604.0    EWR  HNL      4962   
573775             CO         15  N69063    602.0    EWR  HNL      4962   
6982535            CO         14  N76054    523.0    HNL  EWR      4962   
1742577            CO         15  N68061    613.0    EWR  HNL      4962   
5914426            CO         15  N78060    603.0    EWR  HNL      4962   
4199498            CO         15  N66057    610.0    EWR  HNL      4962   
6990216            CO         15  N76055    591.0    EWR  HNL      4962   
4200196            CO         14  N67058    543.0    HNL  EWR      4962   
6977972            CO         15  N59053    613.0    EWR  HNL      4962   
5908596            CO         15  N66056    595.0    EWR  HNL      4962   
4811552            CO         15  N68061    597.0    EWR  HNL      4962   
1742538            CO         14  N69063    552.0    HNL  EWR      4962   
1153923            CO         15  N76055    600.0    EWR  HNL      4962   
1742545            CO         14  N69059    542.0    HNL  EWR      4962   
2969130            CO         15  N59053    604.0    EWR  HNL      4962   
4201674            CO         15  N78060    583.0    EWR  HNL      4962   
2951809            CO         15  N78060    590.0    EWR  HNL      4962   
6976492            CO         15  N68061    624.0    EWR  HNL      4962   
3570353            CO         15  N59053    618.0    EWR  HNL      4962   
4814977            CO         14  N66056    558.0    HNL  EWR      4962   
3561206            CO         15  N76062    612.0    EWR  HNL      4962   
6982536            CO         15  N76054    637.0    EWR  HNL      4962   
2359367            CO         14  N66057    554.0    HNL  EWR      4962   
5910833            CO         15  N66057    595.0    EWR  HNL      4962   
5904165            CO         15  N67052    624.0    EWR  HNL      4962   
5912209            CO         15  N68061    600.0    EWR  HNL      4962   
2362981            CO         15  N68061    613.0    EWR  HNL      4962   
1142728            CO         15  N67052    648.0    EWR  HNL      4962   
1134716            CO         15  N78060    599.0    EWR  HNL      4962   
6428252            CO         15  N67058    603.0    EWR  HNL      4962   
3302108            DL       1273  N831MH    554.0    ATL  HNL      4502   
4228507            DL       1561  N829MH    524.0    ATL  HNL      4502   
4549011            DL       1561  N832MH    532.0    ATL  HNL      4502   
3304566            DL       1273  N834MH    546.0    ATL  HNL      4502   
6212612            DL       1273  N834MH    563.0    ATL  HNL      4502   
2384273            DL       1273  N831MH    543.0    ATL  HNL      4502   
2982547            DL       1282  N833MH    495.0    HNL  ATL      4502   
4222040            DL       1273  N833MH    538.0    ATL  HNL      4502   
6745471            DL       1273  N828MH    549.0    ATL  HNL      4502   
4222052            DL       1282  N833MH    503.0    HNL  ATL      4502   
3591397            DL       1561  N830MH    537.0    ATL  HNL      4502   
5373665            DL       1561  N831MH    541.0    ATL  HNL      4502   
2376989            DL       1273  N830MH    574.0    ATL  HNL      4502   
2098473            DL       1273  N833MH    566.0    ATL  HNL      4502   
4830013            DL       1273  N830MH    516.0    ATL  HNL      4502   

         Velocity  ArrDelay  
5909296      8.12      41.0  
1742551      7.98       0.0  
4808097      8.30      -4.0  
5906279      8.50       2.0  
1147136      8.00      76.0  
17

Podem observar que el mateix vol apareix moltes vegades, ja que al llarg de l'any es duu a terme diferents dies i mesos. Per tal de poder tenir un llistat dels vols més llargs independentment de quantes vegades es duin a terme crearem una variable amb valors únics combinant Origen i Destí.

In [22]:
combinaciones_unicas = taula_vols_ord.drop_duplicates(subset=["Origin", "Dest"])
combinaciones_unicas[["Origin", "Dest", "UniqueCarrier", "FlightNum", "TailNum", "AirTime", "Distance", "Velocity"]]

Origin Dest UniqueCarrier  FlightNum TailNum  AirTime  Distance  \
5909296    EWR  HNL            CO         15  N68061    611.0      4962   
6982535    HNL  EWR            CO         14  N76054    523.0      4962   
3302108    ATL  HNL            DL       1273  N831MH    554.0      4502   
2982547    HNL  ATL            DL       1282  N833MH    495.0      4502   
478093     ORD  HNL            AA         73  N392AA    548.0      4243   
...        ...  ...           ...        ...     ...      ...       ...   
174799     SNA  LAX            OO       5467  N217SW     25.0        36   
177988     LAX  SNA            OO       5471  N216SW     21.0        36   
5319578    PSG  WRG            AS         64  N763AS     10.0        31   
1720062    WRG  PSG            AS         65  N782AS     13.0        31   
4390733    HPN  JFK            OH       5052  N442CA     23.0        30   

         Velocity  
5909296      8.12  
6982535      9.49  
3302108      8.13  
2982547      9.09  
478093       7.74  
...           ...  
174799       1.44  
177988       1.71  
5319578      3.10  
1720062      2.38  
4390733      1.30  

[4812 rows x 8 columns]

Podem observar que hi ha 4812 vols diferents. Els vols més llargs son:
- els vols entre EWR i HNL amb una distància de 4962 milles
- els vols entre ATL i HNL amb una distància de 4502 milles
- els vols entre ORD i HNL amb una distància de 4253 milles

- I els més endarrerits? 

In [23]:
df_delay_o = df_reduced.sort_values(by="ArrDelay", ascending=False)
df_delay_o[["Month","DayofMonth","ArrDelay","Origin", "Dest", "UniqueCarrier", "FlightNum", "TailNum", "AirTime", "Distance", "Velocity"]]

Month  DayofMonth  ArrDelay Origin Dest UniqueCarrier  FlightNum  \
1705852      3           8    1370.0    LGA  MIA            AA       2093   
1006701      2          22    1350.0    PDX  MSP            NW        218   
5784207     10          16    1346.0    SMF  MSP            NW        394   
3435921      6          16    1344.0    LAS  MSP            NW        776   
6916589     12          17    1312.0    LAS  DFW            AA       1138   
...        ...         ...       ...    ...  ...           ...        ...   
3971544      7           3     -48.0    LGA  ATL            FL        371   
6694102     12           1     -50.0    JFK  PHX            US         19   
2969927      5          25     -52.0    EWR  MSY            CO        816   
1451171      3          11     -54.0    PHL  SAN            US       1147   
350426       1          17     -67.0    DEN  ATL            F9        426   

        TailNum  AirTime  Distance  Velocity  
1705852  N3CDAA    177.0      1097      6.20  
1006701  N583NW    169.0      1426      8.44  
5784207  N353NW    184.0      1518      8.25  
3435921  N501US    159.0      1300      8.18  
6916589  N610AA    124.0      1055      8.51  
...         ...      ...       ...       ...  
3971544  N344AT    105.0       761      7.25  
6694102  N656AW    267.0      2153      8.06  
2969927  N14320    148.0      1167      7.89  
1451171  N188US    297.0      2369      7.98  
350426   N922FR     66.0      1199     18.17  

[200000 rows x 11 columns]

Busca les rutes més llargues i les que acumulen més retards.

In [25]:
vols_llarg_ret = taula_vols_ord.groupby(["Origin", "Dest"]).first().reset_index()
vols_llarg_ret_ord = vols_llarg_ret.sort_values(by="ArrDelay", ascending=False)
vols_llarg_ret_ord

Origin Dest UniqueCarrier  FlightNum TailNum  AirTime  Distance  \
354     AZO  MSP            9E       4756  89779E     75.0       426   
4482    SMF  HNL            HA         19  N594HA    310.0      2462   
1539    EGE  MIA            AA       2398  N5CBAA    205.0      1811   
2404    LAS  MIA            AA       1842  N5BVAA    266.0      2175   
1055    DAB  EWR            CO        283  N14613    118.0       884   
...     ...  ...           ...        ...     ...      ...       ...   
2314    JFK  SEA            DL        627  N698DL    317.0      2421   
2949    MHT  LAS            WN        644  N259WN    302.0      2356   
3641    PHL  AUS            WN        125  N335SW    201.0      1430   
3683    PHL  SAN            US       1147  N715UW    295.0      2369   
2322    JFK  STT            DL        503  N706TW    192.0      1623   

      Velocity  ArrDelay  
354       5.68     740.0  
4482      7.94     714.0  
1539      8.83     667.0  
2404      8.18     606.0  
1055      7.49     478.0  
...        ...       ...  
2314      7.64     -29.0  
2949      7.80     -29.0  
3641      7.11     -32.0  
3683      8.03     -33.0  
2322      8.45     -34.0  

[4812 rows x 9 columns]

- Aporta allò que consideris rellevant.

Per tal de fer un anàlisis més exaustiu dels retards. Es crearà un dataset només amb els vols estàn considerats amb retard. 

In [26]:
df_delay = df_reduced[(df_reduced["ArrDelay"] >15)]
print(df_delay.shape)

(126109, 26)


In [27]:
df_delay

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
4474620      8          16          6   2151.0        2050     38.0   
6833490     12           1          1   2019.0        1835   2125.0   
5378051      9          10          3   2200.0        2140   2251.0   
3425312      6           5          4   1011.0        1005   1600.0   
9790         1           6          7   2212.0        2020   2347.0   
...        ...         ...        ...      ...         ...      ...   
124351       1           1          2   2122.0        1950   2302.0   
2970098      5          27          2   1914.0        1725   2350.0   
1127926      2          19          2   2115.0        1945   2239.0   
4482356      8           7          4   1056.0        1021   1232.0   
3707690      7          15          2   2013.0        1940   2141.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ...  Dest  Distance  \
4474620        2337            UA        628  N414UA  ...   DCA       612   
6833490        1930            MQ       3956  N648MQ  ...   GRB       174   
5378051        2229            DL        965  N930DL  ...   STL       483   
3425312        1532            NW        306  N347NW  ...   MSP      1536   
9790           2150            WN       3135  N388SW  ...   OAK       407   
...             ...           ...        ...     ...  ...   ...       ...   
124351         2137            XE       2938  N13956  ...   HSV       595   
2970098        2025            CO       1426  N33203  ...   SAN      2425   
1127926        2110            B6         33  N267JB  ...   JFK       264   
4482356        1154            UA        900  N924UA  ...   SFO       447   
3707690        2113            XE       7771  N16170  ...   LAX       390   

         CarrierDelay  WeatherDelay  NASDelay SecurityDelay LateAircraftDelay  \
4474620          21.0           0.0       0.0           0.0              40.0   
6833490         104.0           0.0      11.0           0.0               0.0   
5378051          12.0           0.0       2.0           0.0               8.0   
3425312           6.0           0.0      22.0           0.0               0.0   
9790              1.0           0.0       5.0           0.0             111.0   
...               ...           ...       ...           ...               ...   
124351            0.0           0.0       0.0           7.0              78.0   
2970098           0.0           0.0     205.0           0.0               0.0   
1127926          89.0           0.0       0.0           0.0               0.0   
4482356           0.0           0.0      38.0           0.0               0.0   
3707690           0.0           0.0       0.0           0.0              28.0   

         Delay  DepDelMin  Velocity  
4474620    Yes      101.0      7.20  
6833490    Yes      184.0      5.61  
5378051    Yes       60.0      5.96  
3425312    Yes        6.0      7.38  
9790       Yes      192.0      5.81  
...        ...        ...       ...  
124351     Yes      172.0      7.44  
2970098    Yes      189.0      6.91  
1127926    Yes      170.0      3.67  
4482356    Yes       35.0      6.21  
3707690    Yes       73.0      5.82  

[126109 rows x 26 columns]

In [28]:
df_reduced["UniqueCarrier"].value_counts()
uniq_carr_del = df_reduced

De 200.000 vols 126.109 han arribat amb retard. El que suposa un 63% de tots els vols.

Per estudiar que ocasiona principalment els retards, es calcularà el total de minuts de retard total i per cada categoria:
- Carr: ocasionats per la companyia aèrea
- Weat: ocasionats per les condicions climatològiques
- NAS: ocasionats per la Agencia Nacioanal Espacial
- Sec: ocasionats per temes relacionats amb la Seguretat
- LAD: ocasionats per avions que venien d'altres destins amb retard.

In [29]:
total_delay = df_delay["ArrDelay"].sum() 
total_delay_carr = df_delay["CarrierDelay"].sum()
total_delay_weat = df_delay["WeatherDelay"].sum()
total_delay_NAS = df_delay["NASDelay"].sum()
total_delay_Sec = df_delay["SecurityDelay"].sum()
total_delay_LAD = df_delay["LateAircraftDelay"].sum()
per_delay_carr = round((total_delay_carr / total_delay) * 100, 2)
per_delay_weat = round((total_delay_weat / total_delay)*100 , 2)
per_delay_NAS = round((total_delay_NAS / total_delay)*100 , 2)
per_delay_Sec = round((total_delay_Sec / total_delay)*100 , 2)
per_delay_LAD = round((total_delay_LAD / total_delay)*100 , 2)

print("Del total de minuts dels retards:{}\n -{}% CARR \n -{}% Weat\n -{}%  NAS\n -{}%  Seg \n -{}% LAD.".format(total_delay, per_delay_carr , per_delay_weat, per_delay_NAS ,
per_delay_Sec , per_delay_LAD))

Del total de minuts dels retards:8159283.0
 -30.3% CARR 
 -5.76% Weat
 -23.85%  NAS
 -0.14%  Seg 
 -39.95% LAD.


Si ho refactoritzem quedaria així:

In [30]:
columns = {
    "CarrierDelay": "CARR",
    "WeatherDelay": "Weat",
    "NASDelay": "NAS",
    "SecurityDelay": "Seg",
    "LateAircraftDelay": "LAD"
}

total_delay = df_delay["ArrDelay"].sum()
frequencies = {}

for column, label in columns.items():
    total_column_delay = df_delay[column].sum()
    percentage_delay = round((total_column_delay / total_delay) * 100, 2)
    frequencies[label] = percentage_delay

print("Del total de minuts dels retards: {}\n".format(total_delay))

for label, percentage in frequencies.items():
    print("- {}% són {}".format(percentage, label))


Del total de minuts dels retards: 8159283.0

- 30.3% són CARR
- 5.76% són Weat
- 23.85% són NAS
- 0.14% són Seg
- 39.95% són LAD


Per saber les  Aerolinies que acumulen més retads:

In [32]:
df_delay["UniqueCarrier"].value_counts()

WN    20587
AA    13622
MQ     9824
UA     9438
OO     9042
DL     7382
XE     7310
US     5923
CO     5814
NW     5607
EV     5572
YV     5198
FL     4740
OH     4089
B6     3871
9E     3683
AS     2368
F9     1542
HA      468
AQ       29
Name: UniqueCarrier, dtype: int64

Per saber els Origens que acumulen més retads:

In [33]:
df_delay["Origin"].value_counts()

ORD    9179
ATL    8842
DFW    6190
DEN    4525
EWR    3674
       ... 
ALO       2
BPT       2
RFD       2
CMX       2
HTS       1
Name: Origin, Length: 296, dtype: int64

Per saber els destins que acumulen més retards:

In [34]:
df_delay["Dest"].value_counts()

ORD    8179
ATL    7161
DFW    4651
EWR    4145
DEN    3804
       ... 
ALO       1
BPT       1
HTS       1
LWB       1
INL       1
Name: Dest, Length: 297, dtype: int64

### Exercici 3

Exporta el dataset net i amb les noves columnes a Excel.

Per tal de mantenir el dataset original, crearem un dataset amb el nom de DelayedFlightupdated.xlsx

In [ ]:
ruta_archivo1 = "C:\\Users\\gemma\\OneDrive\\Escritorio\\It Academy Tasks\\Archivo_ Spr3_aviones\\DelayedFlightsupdated.xlsx"
df_reduced.to_excel(ruta_archivo1 , index=False)